model/system_metrics.go

```diff
@@ -315,3 +368,148 @@ func (id *SystemMetricsInfo) ID() string {
 
 	return fmt.Sprintf("%x", hash.Sum(nil))
 }
+
+// NewSystemMetricsReader returns an io.Reader for the chunks of a particular metric
+// in a provided bucket.
+func NewSystemMetricsReader(ctx context.Context, bucket pail.Bucket, chunks MetricChunks, batchSize int) *SystemMetricsReader {
+	return &SystemMetricsReader{
+		ctx:       ctx,
+		bucket:    bucket,
+		batchSize: batchSize,
+		chunks:    chunks.Chunks,
+		format:    chunks.Format,
+	}
+}
+
+// SystemMetricsReader is a batched, parallelized io.Reader for system metrics data.
+type SystemMetricsReader struct {
+	ctx         context.Context
+	bucket      pail.Bucket
+	batchSize   int
+	chunks      []string
+	chunkIndex  int
+	readerIndex int
+	readers     map[string]io.ReadCloser
+	leftOver    []byte
+	format      FileDataFormat
+	buffer      []byte
+	exhausted   bool
+}
+
+// Format returns the format of the data being read.
+func (s *SystemMetricsReader) Format() FileDataFormat { return s.format }
+
+// Read fills the buffer p with the next data, returning the number of bytes read.
+// If there are no bytes to read, an EOF error is returned.
+func (s *SystemMetricsReader) Read(p []byte) (int, error) {
+	n := 0
+
+	if s.leftOver != nil {
```

Comment:
Can we move this block into a helper function?

Reply:
I don't think its worth it, its only a couple of lines

Reply:
Sure, but the logic is non-obvious what/why it has to do this. Typically, I make a function as the documentation if I'm doing something weird and necessary but non-obvious since it's an easy source of bugs.